OctopusDeploy/OctopusTentacle
# Refactor Delete Instance Wizard in Tentacle Manager

Move some :cheese: in the Tentacle Manager's "Delete Instance" wizard.

# Background
We want to add some automated tests to Tentacle Manager as part of the [Tentacle Manager on .NET](https://app.shortcut.com/octopusdeploy/epic/68113/tentacle-manager-on-windows-net-core) project.

We first need to make the Tentacle Manager project testable by automated tests (e.g. NUnit) by separating the UI logic from the business logic. This will allow the automated tests to exercise most of the code (i.e. everything _not_ in a UI component) without the pain of dealing with UI threads and Dispatchers etc.

[sc-70325]

# Results

There should be *zero visible changes* from an end-user's perspective. The intent is that the functionality is completely unchanged, but the code base is in a better shape for automated tests.

Summary of the main changes within:

#### Remove `ShellView` factory method from container
UI components must be created on the UI thread, which gets real messy when the container is involved. It's best to just not do it, and only use the container for resolving models and other business logic components.

#### `DeleteWizardModel` and `TentacleManagerModel` now derive from `ShellViewModel` 
Existing logic means that multiple models must be created when opening a new Tentacle Manager window - one for the `ShellView`, and one for the `ShellView`'s UI content.

`ShellViewModel` only contains properties that are common to all features being hosted in the `ShellView`, so it makes sense to simplify window creation by allowing `ShellView` to use the same view model as its content.

#### Remove `DeleteWizardLauncher`
This class' only responsibility was to create the models and views required for the "Delete Instance" wizard. Creating models will now be done via the container, and creating the views will be done manually within the UI logic, so the UI creation functionality has been moved to `TentacleManagerView` where it is used.

#### `TentacleManagerModel` creates the `DeleteWizardModel`
It should not be the responsibility of a UI component to know how to create models for a child UI component. `DeleteWizardModel` is now registered in the container, and `TentacleManagerModel` depends on a factory function to create them on demand i.e. whenever the Tentacle Manager view wants to open a new instance of the "Delete Instance" wizard.

#### `TentacleManagerView` depends on the container
While the end goal is to have views relatively ignorant of how to create models, iteratively implementing this means we need to let `TentacleManagerView` create/resolve components it probably shouldn't be... at least until we've converted all the views/components.

#### Add exclusion for local dev builds to the `IsEarlyAccessProgram` helper method
By removing some of the usages of the container to resolve `ShellViewModel`, a side effect was introduced. [This line, which has been removed,](https://github.com/OctopusDeploy/OctopusTentacle/compare/sast/tentacle-manager/delete-wizard?expand=1#diff-77d439cef0dfaaf630696e22466d1bc6b423d232a96943d3f0bdace00c5f73acL21) added a handler action for the container's OnActivating event when resolving a `ShellViewModel`. This action modified the resolved instance by disabling the EAP message. While it is highly unlikely we will be releasing any EAP versions of Tentacle Manager in the future, rather than delete this functionality it has been updated so that the EAP message gets disabled if the build is a local dev build i.e. Major version starts with 0, such as `0.0.0-local`.

# How to review this PR

Quality :heavy_check_mark:

As there are no automated tests, the only way to verify this PR is to manually test the Tentacle Manager. There's not much functionality in here, so testing the following should cover everything:
* Install new instance
* Reinstall existing instance
* Restart service
* Start service
* Stop service
* Update proxy settings
* Delete instance

# Pre-requisites

- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
